tests/unit/package_managers/yarn_classic/test_utils.py

```diff
@@ -0,0 +1,102 @@
+from unittest import mock
+
+from cachi2.core.package_managers.yarn_classic.project import PackageJson
+from cachi2.core.package_managers.yarn_classic.utils import find_runtime_deps
+from cachi2.core.rooted_path import RootedPath
+
+PACKAGE_JSON = """
+{
+  "name": "project",
+  "dependencies": {
+    "main-dep1": "^1.0.0"
+  },
+  "optionalDependencies": {
+    "optional-dep1": "^2.0.0"
+  },
+  "peerDependencies": {
+    "peer-dep1": "^3.0.0"
+  },
+  "devDependencies": {
+    "dev-dep1": "^4.0.0"
+  }
+}
+"""
+
+
+@mock.patch("cachi2.core.package_managers.yarn_classic.project.YarnLock")
+@mock.patch("cachi2.core.package_managers.yarn_classic.utils._expand_yarn_lock_keys")
+def test_find_runtime_deps(
+    mock_expand_yarn_lock_keys: mock.Mock,
+    mock_yarn_lock: mock.Mock,
+    rooted_tmp_path: RootedPath,
+) -> None:
+    rooted_tmp_path.join_within_root("package.json").path.write_text(PACKAGE_JSON)
+    package_json = PackageJson.from_file(rooted_tmp_path.join_within_root("package.json"))
+
+    mock_expand_yarn_lock_keys.return_value = {
+        # dependencies
+        "main-dep1@^1.0.0": {
+            "version": "1.0.0",
+            "dependencies": {"sub-dep1": "^1.1.0"},
+        },
+        # optional dependencies
+        "optional-dep1@^2.0.0": {
+            "version": "2.0.0",
+            "dependencies": {"multi-dep1": "^4.0.0", "multi-dep2": "^4.0.0"},
+        },
+        "multi-dep1@^4.0.0, multi-dep2@^4.0.0": {"version": "5.0.0", "dependencies": {}},
+        # peer dependencies
+        "peer-dep1@^3.0.0": {"version": "3.0.0", "dependencies": {}},
+        # transitive dependencies
+        "sub-dep1@^1.1.0": {"version": "1.1.0", "dependencies": {"sub-dep11": "^1.1.1"}},
+        "sub-dep11@^1.1.1": {"version": "1.1.1", "dependencies": {}},
+        # dev dependencies
+        "dev-dep1@^4.0.0": {"version": "4.0.0", "dependencies": {}},
+    }
+
+    result = find_runtime_deps(package_json, mock_yarn_lock, [])
+
+    mock_expand_yarn_lock_keys.assert_called_once_with(mock_yarn_lock)
+    assert result == {
+        "main-dep1@^1.0.0",
+        "optional-dep1@^2.0.0",
+        "multi-dep1@^4.0.0",
+        "multi-dep2@^4.0.0",
+        "peer-dep1@^3.0.0",
+        "sub-dep1@^1.1.0",
+        "sub-dep11@^1.1.1",
+        # no dev dependencies
+    }
+
+
+PACKAGE_JSON_WITH_CONFLICT = """
+{
+  "name": "project",
+  "dependencies": {
+    "foo": "^1.1.1"
+  },
+  "devDependencies": {
+    "foo": "^1.1.1"
+  }
+}
+"""
+
+
+@mock.patch("cachi2.core.package_managers.yarn_classic.project.YarnLock")
+@mock.patch("cachi2.core.package_managers.yarn_classic.utils._expand_yarn_lock_keys")
+def test_find_runtime_deps_with_conflict_in_dev_dependencies(
```

Comment:
I don't think workspaces are needed here, all package.json files are iterated without any differentiation.

> Will it work in case when a dev dependency becomes a runtime dependency transitively?

What do you mean by that ?

Reply:
Sorry, I had to be more explicit with my example. I meant that in a hypothetical test `MAIN_PJ` should be passed as main yarn lock and SUBPROJECT_PJ as a workspace yarn lock. In the example above my expectation is that `bar` should get promoted to runtime dependency.

Reply:
> In the example above my expectation is that bar should get promoted to runtime dependency.

That's true, but there is only one `yarn.lock` in the repo that contains keys for all dependencies.

Reply:
My bad, `package.json`, not yarn lock, calling thing by their actual names is hard. `find_runtime_deps` takes main `package.json` a lock, and a list of possible other `package.json`s for nested packages. I am interested in a situation when `bar` is declared as a devDependency in main `package.json` and as a runtime dependency in one of workspaces.

Reply:
Alright, I added a test with workspace. 

> I am interested in a situation when bar is declared as a devDependency in main package.json and as a runtime dependency in one of workspaces.

In that case, `yarn.lock` would contain one entry for that dependency. When iterating over `package.json` from the workspace, the dependency would be found inside the `dependencies` dictionary, so it would be marked as runtime. Dev dependencies are kinds of leftover records that were not identified as runtime in any `package.json` or as transitive dependencies.

Reply:
Awesome, thank you!